base/iostream.jl

```diff
@@ -443,11 +443,46 @@ end
 function readuntil_string(s::IOStream, delim::UInt8, keep::Bool)
     @_lock_ios s ccall(:jl_readuntil, Ref{String}, (Ptr{Cvoid}, UInt8, UInt8, UInt8), s.ios, delim, 1, !keep)
 end
+readuntil(s::IOStream, delim::AbstractChar; keep::Bool=false) =
+    delim ≤ '\x7f' ? readuntil_string(s, delim % UInt8, keep) :
+    String(unsafe_take!(copyuntil(IOBuffer(sizehint=70), s, delim; keep)))
 
 function readline(s::IOStream; keep::Bool=false)
     @_lock_ios s ccall(:jl_readuntil, Ref{String}, (Ptr{Cvoid}, UInt8, UInt8, UInt8), s.ios, '\n', 1, keep ? 0 : 2)
 end
 
+function copyuntil(out::IOBuffer, s::IOStream, delim::UInt8; keep::Bool=false)
+    ensureroom(out, 16)
+    ptr = (out.append ? out.size+1 : out.ptr)
+    d = out.data
+    len = length(d)
+    while true
+        GC.@preserve d @_lock_ios s n=
+            Int(ccall(:jl_readuntil_buf, Csize_t, (Ptr{Cvoid}, UInt8, Ptr{UInt8}, Csize_t),
+                s.ios, delim, pointer(d, ptr), (len - ptr + 1) % Csize_t))
+        iszero(n) && break
+        ptr += n
+        if d[ptr-1] == delim
+            keep || (ptr -= 1)
+            break
+        end
+        (eof(s) || len == out.maxsize) && break
+        len = min(2len + 64, out.maxsize)
+        resize!(d, len)
```

Comment:
This seems like it might not quite respect maxsize. Should it use `ensureroom` instead? The `ensureroom` call at the top also seems wrong, since if someone new the `out` stream was already sized correctly for a smaller read, that will forceably try to reallocate it.

Reply:
I'm not sure what you mean by "not copying the last chunk".  By the time it hits the `len == out.maxsize` branch, `jl_readuntil_buf` has already copied `out.maxsize` bytes into the buffer.   The `jl_readuntil_buf` never reads a chunk exceeding `maxsize`, because the number of bytes that it reads is bounded by `len - ptr + 1`.

`ensureroom` seems like it has quite a few extra checks that aren't needed in the inner loop here given the `ensureroom` call at the top.

I guess it wouldn't hurt to change the `ensureroom` call at the top to something like `ensureroom(out, iszero(io.size) ? 16 : 0)`

Reply:
Ah, I guess that makes sense. The checks were a bit distant so it wasn't immediately obvious that was being ensured.

Yes, I think checking `isempty` first makes sense. I realized now that `ensureroom` already truncates the request to maxsize, so there is no issue with calling that.

Reply:
should be resolved by #50485